engines/leveldb/leveldb_test.go

```diff
@@ -0,0 +1,235 @@
+}
+
+func TestLeveldbEngine(t *testing.T) {
+	// Setup a levelDB Engine.
+	ldbEngine, ldbPath := setupLeveldbEngine(t)
+	require.Equal(t, "leveldb", ldbEngine.Engine())
+
+	// Ensure the db was created.
+	require.DirExists(t, ldbPath)
+
+	// Teardown after finishing the test.
+	defer os.RemoveAll(ldbPath)
+	defer ldbEngine.Close()
+
+	// Use a constant key to ensure namespaces
+	// are working correctly.
+	key := []byte("foo")
+
+	// Check Put, Get and Delete with a nil namespace.
+	value := []byte("nil")
+	checkPut(ldbEngine, nil, key, value, t)
+	checkGet(ldbEngine, nil, key, value, t)
+	checkDelete(ldbEngine, nil, key, t)
+
+	// Iterate through a list of namespaces and ensure
+	// Put, Get and Delete are working.
+	for _, namespace := range getNamespaces() {
+		opts := namespaceOpts(namespace, t)
+		value := []byte(namespace)
+		checkPut(ldbEngine, opts, key, value, t)
+		checkGet(ldbEngine, opts, key, value, t)
+		checkDelete(ldbEngine, opts, key, t)
+	}
+}
+
+func TestLeveldbTransactions(t *testing.T) {
+	ldbEngine, ldbPath := setupLeveldbEngine(t)
+
+	// Teardown after finishing the test
+	defer os.RemoveAll(ldbPath)
+	defer ldbEngine.Close()
+
+	// Use a constant key to ensure namespaces
+	// are working correctly.
+	key := []byte("tx")
+
+	// Iterate through a list of namespaces and ensure
+	// Put, Get and Delete are working with transactions.
+	for _, namespace := range getNamespaces() {
+		// Start a transaction with readonly set to false.
+		tx, err := ldbEngine.Begin(false)
+		require.NoError(t, err)
+
+		opts, err := options.New(options.WithNamespace(namespace))
+		require.NoError(t, err)
+		value := []byte(namespace)
+		checkPut(tx, opts, key, value, t)
+		checkGet(tx, opts, key, value, t)
+		checkDelete(tx, opts, key, t)
+
+		// Complete the transaction.
+		require.NoError(t, tx.Finish())
+	}
+
+	// Begin a transaction with readonly set to true.
+	tx, err := ldbEngine.Begin(true)
+	require.NoError(t, err)
+
+	// Ensure Put fails with readonly set.
+	err = tx.Put([]byte("bad"), []byte("write"), nil)
+	require.Equal(t, err, engine.ErrReadOnlyTx)
+	err = tx.Delete([]byte("bad"), nil)
+	require.Equal(t, err, engine.ErrReadOnlyTx)
+
+	// Complete the transaction.
+	require.NoError(t, tx.Finish())
+}
+
+func TestLevelDBIter(t *testing.T) {
+	ldbEngine, ldbPath := setupLeveldbEngine(t)
+
+	// Teardown after finishing the test
+	defer os.RemoveAll(ldbPath)
+	defer ldbEngine.Close()
+
+	for _, namespace := range getNamespaces() {
+		// TODO: figure out what to do with this testcase.
+		// Iter currently grabs the namespace by splitting
+		// on :: and grabbing the first string, so it only
+		// grabs "namespace".
+		if namespace == "namespace::with::colons" {
+			continue
+		}
+		// Add data to the database to iterate over.
+		opts := namespaceOpts(namespace, t)
+		pairs := [][]string{
+			{"aa", "first"},
+			{"ab", "second"},
+			{"ba", "third"},
+			{"bb", "fourth"},
+			{"bc", "fifth"},
+			{"ca", "sixth"},
+			{"cb", "seventh"},
+		}
```

Comment:
Since this is being passed around to several functions, it's probably a good candidate to be moved to a top-level variable, e.g. as a fixture that's used by multiple functions - particularly since it's not being modified, just read.

Reply:
changed